backend/src/database/state-migrations/migrations.ts

```diff
@@ -52,55 +54,59 @@ export const migrations: {
+ */
+export async function migrateInDatabase(
+    targetStateVersion: number | 'latest',
     exerciseId: UUID,
     entityManager: EntityManager
 ): Promise<void> {
-    const migrationFunctions = getMigrationFunctions(
-        currentStateVersion,
-        targetStateVersion
-    );
-    const stateMigrationFunctions = migrationFunctions.state;
-    const originalStates = await entityManager.findOne(ExerciseWrapperEntity, {
+    const exercise = await entityManager.findOne(ExerciseWrapperEntity, {
         where: { id: exerciseId },
-        select: { initialStateString: true, currentStateString: true },
     });
-    if (originalStates === null) {
+    if (exercise === null) {
         throw new RestoreError(
             'Cannot find exercise to convert in database',
             exerciseId
         );
     }
-    let initialState: object | null = null,
-        currentState: object | null = null;
-    if (stateMigrationFunctions.length > 0) {
-        initialState = JSON.parse(originalStates.initialStateString);
-        currentState = JSON.parse(originalStates.currentStateString);
-        stateMigrationFunctions.forEach((stateMigrationFunction) => {
-            initialState = stateMigrationFunction(initialState!);
-            currentState = stateMigrationFunction(currentState!);
-        });
-    }
-    const actionsMigrationFunctions = migrationFunctions.actions;
-    let actions: (object | null)[] | null = null;
-    if (actionsMigrationFunctions.length > 0) {
-        const originalActions = await entityManager.find(ActionWrapperEntity, {
+    const targetVersion =
+        typeof targetStateVersion === 'number'
+            ? targetStateVersion
+            : Math.max(...StrictObject.keys(migrations));
+    if (targetStateVersion <= exercise.stateVersion) return;
+    let initialState = JSON.parse(exercise.initialStateString);
+    let currentState = JSON.parse(exercise.currentStateString);
+    const actions = (
+        await entityManager.find(ActionWrapperEntity, {
             where: { exercise: { id: exerciseId } },
             select: { actionString: true },
             order: { index: 'ASC' },
-        });
-        actions = originalActions.map((originalAction) =>
-            JSON.parse(originalAction.actionString)
-        );
-        initialState ??= JSON.parse(originalStates.initialStateString);
-        actionsMigrationFunctions.forEach((actionsMigrationFunction) => {
-            actionsMigrationFunction(initialState!, actions!);
-        });
+        })
+    ).map((action) => JSON.parse(action.actionString));
+    for (
+        let currentStateVersion = exercise.stateVersion + 1;
+        currentStateVersion <= targetVersion;
+        currentStateVersion++
+    ) {
+        const stateMigration = migrations[currentStateVersion]?.state;
+        if (typeof stateMigration === 'function') {
+            initialState = stateMigration(initialState);
+            currentState = stateMigration(currentState);
```

Comment:
I believe this will create problems in the future.

It should always be that `applyAllActions(initialState, actions) == currentState`.
In the latest migrations in #523 we had to create a SpatialTree. The specific SpatialTree depends on the order of the actions that were performed. Therefore `applyAllActions(initialState, actions) == currentState` could be false. Besides failing checks with `isEqual()` (what could be very confusing) it also effects the state itself as the treatment algorithm could potentially have a different result if e.g. two patients have the same distance from a personnel and, due to them changing the order in the spatial tree, the personnel now treats someone else.

I would therefore recommend to first, migrate the `initialState` and all the actions to the wanted version, and then applying all the actions to get the `currentState`. If there is an Error during the application of the actions, we know that the migration is incorrect and this exercise could not be migrated.

Reply:
In `shared\src\store\time-travel-utilities.ts`
```ts
/**
 * @param actions all the actions that were applied to the {@link initialState} to get to the present state
 * @returns the state after applying all the actions
 */
export function applyAllActions(
    initialState: ExerciseState,
    actions: readonly ExerciseAction[]
): ExerciseState {
    // TODO: find a heuristic for whether using produce or deepCloning it
    // Default is produce
    return produce(initialState, (draftState) => {
        for (const action of actions) {
            applyAction(draftState, action);
        }
    });
}
```

Reply:
Hm, good point. Although I didn't change it in this PR, I believe it was incorrect even before (but the function was mich more complex so it was more difficult to see). I will change it to only migrate the `initialState` and the `actions` and then use them to create a new `currentState` (unless we have a `StateExport` without `history`).